terraso_backend/apps/audit_logs/services.py

```diff
@@ -0,0 +1,213 @@
+import typing
+from datetime import datetime
+from typing import List, Optional
+
+from django.contrib.contenttypes.models import ContentType
+from django.core.paginator import Paginator
+from django.db import transaction
+from django.db.models.query import QuerySet
+
+from . import api, models
+
+TEMPLATE = "{client_time} - {user} {action} {resource}"
+
+
+class _AuditLogService:
+    """
+    AuditLogService implements the AuditLog protocol
+    """
+
+    def log(
+        self,
+        user: object,
+        action: int,
+        resource: object,
+        metadata: typing.Optional[List[api.KeyValue]] = None,
+    ) -> None:
+        """
+        log logs an action performed by a user on a resource
+        example:
+            log(user, "create", resource, [("client_time", 1234567890)])
+            :param metadata:
+            :param action:
+            :param user:
+            :type resource: object
+        """
+        if not hasattr(user, "id"):
+            raise ValueError("Invalid user")
+
+        get_user_readable = getattr(user, "human_readable", None)
+        user_readable = get_user_readable() if callable(get_user_readable) else user.id
```

Comment:
I discussed this with Lilian in a call, and we agreed on email, but I can ask on the call tomorrow, Im agnostic to it. That is more of a product question. (The original though was is valid to have duplicated names)

Reply:
@ltseng, what do you think about this?

Reply:
I had said email at the time because I was under the impression that was less likely to change for any given user, and less likely to have duplicates. If name turns out to be easier to accomplish, that is fine as well.

Reply:
Either is the same level of effort. I can do whatever you think is best.